src/main/java/org/apache/mesos/scheduler/plan/DefaultPhaseFactory.java

```diff
@@ -1,38 +1,68 @@
 package org.apache.mesos.scheduler.plan;
 
-import org.apache.mesos.offer.InvalidRequirementException;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.mesos.scheduler.plan.strategy.SerialStrategy;
+import org.apache.mesos.scheduler.plan.strategy.Strategy;
+import org.apache.mesos.scheduler.plan.strategy.StrategyGenerator;
 import org.apache.mesos.specification.TaskSet;
-import org.apache.mesos.specification.TaskSpecification;
 
-import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
-import java.util.UUID;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 /**
  * This class generates Phases given PhaseSpecifications.
  */
-public class DefaultPhaseFactory {
+public class DefaultPhaseFactory implements PhaseFactory {
 
     private final BlockFactory blockFactory;
+    private final StrategyGenerator<Block> strategyGenerator;
 
     public DefaultPhaseFactory(BlockFactory blockFactory) {
-       this.blockFactory = blockFactory;
+        this(blockFactory, new SerialStrategy.Generator<>());
     }
 
-    public Phase getPhase(TaskSet taskSet) throws InvalidRequirementException {
-        return DefaultPhase.create(
-                UUID.randomUUID(),
-                taskSet.getName(),
-                getBlocks(taskSet));
+    public DefaultPhaseFactory(BlockFactory blockFactory, StrategyGenerator<Block> strategyGenerator) {
+        this.blockFactory = blockFactory;
+        this.strategyGenerator = strategyGenerator;
+    }
+
+    @Override
+    public Phase getPhase(TaskSet taskSet) {
+        return getPhase(taskSet, strategyGenerator.generate());
     }
 
-    private List<Block> getBlocks(TaskSet taskSet) throws InvalidRequirementException {
-        List<Block> blocks = new ArrayList<>();
+    @Override
+    public Phase getPhase(TaskSet taskSet, Strategy strategy) {
+        return new DefaultPhase(
+                taskSet.getName(),
+                getBlocks(taskSet),
+                strategy,
+                Collections.emptyList());
+    }
 
-        for (TaskSpecification taskSpecification : taskSet.getTaskSpecifications()) {
-            blocks.add(blockFactory.getBlock(taskSet.getName(), taskSpecification));
-        }
+    @Override
+    public List<Phase> getPhases(List<TaskSet> taskSets, StrategyGenerator strategyGenerator) {
+        return taskSets.stream()
+                .map(taskSet -> getPhase(taskSet, strategyGenerator.generate()))
+                .collect(Collectors.toList());
+    }
 
-        return blocks;
+    private List<Block> getBlocks(TaskSet taskSet) {
+        return taskSet.getTaskSpecifications().stream()
+                .map(taskSpec -> {
+                    try {
+                        return blockFactory.getBlock(taskSpec);
+                    } catch (Block.InvalidBlockException e) {
+                        return new DefaultBlock(
```

Comment:
Just curious about implication of this. If there was a failure in getting a block from a taskSpec it should just fail, right ? Or the implication here is that if there's any block that's in `ERROR` state, the plan will be in error state (status propagates upwards) and will not make any progress, but it will be used to tell the operator via API that there's an error ?